src/Runtime/Values.hs

```diff
@@ -15,13 +15,56 @@ import Data.Array
 import Data.List
 import Data.Aeson hiding (Array)
 import GHC.Generics
+import qualified Data.Map.Strict as Map
 
 -- | Representation of a Board in BoGL,
 -- composed of an NxM array of 'Val'
 type Board = Array (Int, Int) Val
 
--- | Evaluation environment
-type EvalEnv = [(Name, Val)]
+
+-- | Typeclass for a runtime environment
+class RuntimeEnv a where
+  -- insert single key/val pair into the env
+  insertEvalEnv :: (String,Val) -> a -> a
+  -- extend an env with a tuple of keys and vals to be combined
+  extendEvalEnv :: ([String],[Val]) -> a -> a
+  -- combine two environments, producing a new env
+  unionEvalEnv  :: a -> a -> a
+  -- looks up a value in an env
+  lookupEvalEnv :: String -> a -> Maybe Val
+
+
+-- | Different kinds of evaluation environments
+data EvalEnv = ListEvalEnv [(String,Val)] -- env represented with a list
```

Comment:
Now that the intermediate step is done, I think this data type should be removed and instead the `RuntimeEnv` typeclass should be used in signatures instead.